service/src/main/kotlin/fi/espoo/evaka/messaging/MessageQueries.kt

```diff
@@ -389,7 +401,8 @@ WHERE
         } else {
             "tp.folder_id = :folderId"
         }
-    }
+    } AND
+    EXISTS (SELECT 1 FROM message m WHERE m.thread_id = t.id AND m.sent_at < :undoThreshold)
```

Comment:
Have you checked this change with EXPLAIN in prod? It seems to me that this has the potential to slow down the query by a lot.

Reply:
If there is a significant performance hit, it could possibly be mitigated by setting `thread.last_received_timestamp` to be send time + 11 seconds, and then instead filter here that `thread.last_received_timestamp <= :now`, but not sure how hard that is in the sending and undoing code.

Reply:
Hmm, now that I think of it, how well do the the `message_thread_participant.last_*_timestamp` columns now work with this change? If a reply is sent and then undone, does the thread show as the newest thread in the receiver's inbox?

Reply:
I completely forgot about `message_thread_participant`. I'll need to fix that.

When it comes to performance I'm not sure how to test this beforehand. Querying a random person's messages seems pretty questionable.

Reply:
I've now added a function that resets `message_thread_participant` to its previous state when a message is undone.

Reply:
> When it comes to performance I'm not sure how to test this beforehand. Querying a random person's messages seems pretty questionable.

`EXPLAIN ANALYZE` in the production database. It doesn't show you the query result, only the query plan and timing info.